File: python/example6_1.py
"""
# Example 6.1
# One-Variable Scalar-Valued Function
# References: based on STA-663-2017 1.0 Documentation
# http://people.duke.edu/~ccc14/sta-663-2017/14C_Optimization_In_Python.html
# https://docs.scipy.org/doc/scipy/reference/optimize.html
"""

import numpy as np 
import scipy.optimize as opt # from scipy import optimize
import matplotlib.pyplot as plt

# def f(x): return np.log(x)-x**2
# using one-line function
f = lambda x: np.log(x)-x**2
v=0.5
f(v)

# negative of f(x) and its derivative
# def nf(x): return -1*f(x)
# def nf1(x): return 2*x - 1/x
nf = lambda x: -1*f(x)
nf1 = lambda x: 2*x - 1/x
nf2 = lambda x: 2 + 1/(x**2)
nf(v)
nf1(v)
nf2(v)

# using numeric derivatives
# pip install numdifftools
import numdifftools as nd
nd.Gradient(nf,method='complex')(v)
nd.Hessian(nf,method='complex')(v)

# res1 = opt.minimize_scalar(nf)  # method='brent'
res1 = opt.minimize(nf,v,method='bfgs')
res1.fun
res1.x
res1.jac
res1.hess_inv
res2 = opt.minimize(nf,v,jac=nf1,method='newton-cg')
res2

x = np.linspace(1.0e-6,4.0,100)
plt.xlabel('x')
plt.ylabel('f(x)')
plt.scatter(res2.x,-res2.fun)
plt.plot(x,f(x))
plt.show()

fig, (ax1,ax2) = plt.subplots(2,sharex=True)
ax1.set_ylabel('f',rotation=0)
ax2.set_ylabel('df',rotation=0)
ax2.set_ylim(-10,10)
ax1.plot(x,f(x))
ax2.plot(x,-nf1(x))
ax1.scatter(res2.x,-res2.fun)
ax1.vlines(res2.x,ymin=-15,ymax=-res2.fun,linestyles='dotted')
ax2.vlines(res2.x,ymin=-10,ymax=10,linestyles='dotted')
ax2.hlines(0,xmin=0,xmax=4,linestyles='dotted')
plt.show()

File: R/example6_1.R
# Example 6.1
# One-Variable Scalar-Valued Function

setwd("C:/Course17/ceR/R")

f<-function(x) log(x)-x^2  # 1-variable scalar-valued function
v<-0.5          			     # function evaluation at point v=0.5
f(v)                       # f(0.5)

# Numeric Derivatives
library(numDeriv)          # need to install package numDeriv
grad(f,v)                  # 1st numeric derivative of f(x)
hessian(f,v)               # 2nd numeric derivative of f(x)

# Using graph to find the maximum of f
x<-seq(0.0001,4,length.out=40)
y<-f(x)
plot(x,y)
plot(x,y,type="l",ylim=c(-15,0),xlim=c(0,4))

# find maximum of f(x) at x=sqrt(0.5)
max1<-optim(v,f,method="BFGS",control=list(fnscale=-1))
max1$par
max1$value
points(x=max1$par,y=max1$value,col="red",pch=20)

# or, using nlm
fn<-function(x) -f(x)      # negative f, to find minimum of fn
nlm(fn,v)

rm(list=ls())

